src/storage-types/src/connections.rs

```diff
@@ -452,17 +458,187 @@ impl<C: ConnectionAccess> KafkaConnection<C> {
     }
 }
 
+/// A wrapper for use around `rdkafka` types to ensure that they are
+/// safely used within async context. In particular, this ensures that
+/// the _dropping_ of rdkafka clients happens in a separate thread, so that
+/// this does not block the tokio task.
+pub struct RdkafkaWrapper<T: Send + 'static> {
+    /// The inner is an `Option` so it can be taken by the `Drop` impl.
+    ///
+    /// Its an Arc because some `rdkafka` apis take `self: &Arc<Self>`. This
+    /// causes double-indirection in some cases.
+    inner: Option<Arc<T>>,
+}
+
+impl<T: Send + 'static> Drop for RdkafkaWrapper<T> {
+    fn drop(&mut self) {
+        // TODO(guswynn): we would prefer to use `spawn_blocking` here, but we are unsure if all
+        // places this is used has access to the tokio runtime. So, we error log in that case,
+        // letting `sentry` tell us where if any exist.
+        if tokio::runtime::Handle::try_current().is_err() {
+            tracing::error!("dropped rdkafka object from non-tokio context");
+        }
+
+        let inner = self.inner.take().unwrap();
+
+        // Note that `Arc::into_inner` is guaranteed to return `Some(_)` for _exactly_
+        // one clone of the `Arc`.
+        if let Some(inner) = Arc::into_inner(inner) {
+            std::thread::spawn(move || drop(inner));
+        }
+    }
+}
+
+impl<T: Send + 'static> Clone for RdkafkaWrapper<T> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+pub struct RdkafkaWrapperPartitionQueue<C: ConsumerContext + Send + 'static> {
+    // the order of declaration of these values matters. We want to drop
+    // the `PartitionQueue`'s, before the `RdkafkaWrapper`-d consumer, so that the `Drop`
+    // impl of the wrapper runs around the consumer actually sees the final owner of the
+    // value and destructs it in a different thread.
+    partition_queue: rdkafka::consumer::base_consumer::PartitionQueue<C>,
+    _consumer: RdkafkaWrapper<BaseConsumer<C>>,
+}
+
+impl<C: ConsumerContext + Send + 'static> RdkafkaWrapper<BaseConsumer<C>> {
+    /// `split_partition_queue_for_sources` takes `self: &Arc<BaseConsumer>`, but we do NOT want to
+    /// expose access to the inner `Arc` to the user. This would allow them to clone it, and drop
+    /// it without going through `RdkafkaWrapper::drop`, so we wrap that impl here.
+    ///
+    /// Note that the returned `PartitionQueue` is also wrapped, because it contains a clone of its
+    /// `BaseConsumer`.
+    ///
+    // TODO(guswynn): this is probably still unsafe; A `RdkafkaWrapper<PartitionQueue>` that
+    // outlives the original `BaseConsumer` it came from will drop the `BaseConsumer` on the
+    // current thread. This scenario is likely rare, might already cause a deadlock in `rdkafka`,
+    // and requires changing the `rdkafka` crate to resolve. We should figure out how to fix this.
+    pub fn split_partition_queue_for_sources<F>(
```

Comment:
No, because we need to wrap the `PartitionQueue`, and this takes `Arc<Self>`; we don't want to deref to an `Arc<Self>` to make this callable; that would allow the user to clone the underlying consumer

Reply:
I didn't follow that, what's the problem with this:

```diff
diff --git a/src/storage-types/src/connections.rs b/src/storage-types/src/connections.rs
index 6c6a6dcc6d..4a99d5f90d 100644
--- a/src/storage-types/src/connections.rs
+++ b/src/storage-types/src/connections.rs
@@ -518,30 +518,31 @@ impl<C: ConsumerContext + Send + 'static> RdkafkaWrapper<BaseConsumer<C>> {
     // outlives the original `BaseConsumer` it came from will drop the `BaseConsumer` on the
     // current thread. This scenario is likely rare, might already cause a deadlock in `rdkafka`,
     // and requires changing the `rdkafka` crate to resolve. We should figure out how to fix this.
-    pub fn split_partition_queue_for_sources<F>(
+    pub fn split_partition_queue(
         &self,
         topic: &str,
         partition: i32,
-        non_empty_callback: F,
-    ) -> Option<RdkafkaWrapperPartitionQueue<C>>
-    where
-        F: Fn() + Send + Sync + 'static,
-    {
+    ) -> Option<RdkafkaWrapperPartitionQueue<C>> {
         let consumer = self.clone();
         self.inner
             .as_ref()
             .unwrap()
             .split_partition_queue(topic, partition)
-            .map(|mut q| {
-                q.set_nonempty_callback(non_empty_callback);
-
-                RdkafkaWrapperPartitionQueue {
-                    partition_queue: q,
-                    _consumer: consumer,
-                }
+            .map(|q| RdkafkaWrapperPartitionQueue {
+                partition_queue: q,
+                _consumer: consumer,
             })
     }
 }
+impl<C: ConsumerContext + Send + 'static> RdkafkaWrapperPartitionQueue<C> {
+    pub fn set_nonempty_callback<F>(&mut self, non_empty_callback: F)
+    where
+        F: Fn() + Send + Sync + 'static,
+    {
+        self.partition_queue
+            .set_nonempty_callback(non_empty_callback);
+    }
+}
 
 // We don't implement `Deref` generically over any `T`. This is because some `rdkafka` apis
 // (namely, `BaseProducer`) are `: Clone`. Exposing those by `Deref`-ing them would allow the user
diff --git a/src/storage/src/source/kafka.rs b/src/storage/src/source/kafka.rs
index 07347373e9..4b72625c53 100644
--- a/src/storage/src/source/kafka.rs
+++ b/src/storage/src/source/kafka.rs
@@ -776,19 +776,19 @@ impl KafkaSourceReader {
         for pc in &mut self.partition_consumers {
             pc.partition_queue = self
                 .consumer
-                .split_partition_queue_for_sources(&self.topic_name, pc.pid, {
-                    let context = Arc::clone(&context);
-                    move || context.inner().activate()
-                })
+                .split_partition_queue(&self.topic_name, pc.pid)
                 .expect("partition known to be valid");
+            let context = Arc::clone(&context);
+            pc.partition_queue
+                .set_nonempty_callback(move || context.inner().activate());
         }
 
-        let partition_queue = self
+        let mut partition_queue = self
             .consumer
-            .split_partition_queue_for_sources(&self.topic_name, partition_id, move || {
-                context.inner().activate()
-            })
+            .split_partition_queue(&self.topic_name, partition_id)
             .expect("partition known to be valid");
+        partition_queue.set_nonempty_callback(move || context.inner().activate());
+
         self.partition_consumers.push(PartitionConsumer::new(
             partition_id,
             partition_queue,

```